### DISCUSSION TITLE
robotology-superbuild source repositories directories will change in the next weeks

### DISCUSSION THREAD

**Discussion by traversaro** [robotology/community#451] [Announcements]
## Target Users

If you are using the robotology-superbuild only for installing software, for example by using the robotology-superbuild releases, this change will not impact you. Furthermore, this change will only impact releases of the superbuild that will be done in 2021. 

If instead you are using the robotology-superbuild in [rolling update](https://github.com/robotology/robotology-superbuild#rolling-update) workflow and you modify the source code of the robotology-superbuild packages, please ready in the following. 

## Old Behavior 
If you are a user of the `robotology-superbuild`, you know that once you run the superbuild, as part of the build process all the git repositories for the superbuild packages will be cloned in two directories:
* `robotology-superbuild/external` In this directory, packages **not developed** in IIT or part of the robotology organization are contained 
* `robotology-superbuild/robotology` In this directory, packages **that are developed** in IIT or part of the robotology organization are contained.

## New Behavior 
As discussed in https://github.com/robotology/robotology-superbuild/issues/333, this will change in PR https://github.com/robotology/robotology-superbuild/pull/556, that will instead ensure that all repos will be cloned in `robotology-superbuild/src`, regardless of their type. 

## Rationale
The rationale of this change is discussed in https://github.com/robotology/robotology-superbuild/issues/333 , but it is reported in the following:

* Consistency: all the source repos will go in the same directory, and with a directory whose meaning is clear from its name. 
* Familiarity for people used to ROS/ROS2/colcon: in ROS/ROS2 workspaces, all the repos are contained in the `src` subdirectory of the  workspace 
* We can use more easily the `vcs` tool (see https://github.com/dirk-thomas/vcstool)  to manipulate the repos if they are in devel model, for example to export the value of all the repos in the superbuild. `vcs` is the tool used in ROS2, so it would be something familiar used to ROS2 workspaces. In theory `vcs` could be used also with the current directory structure, but the saved file would have the strange format of having the `external` or `robotology` directory prefixed to the CMake package name: 
~~~
repositories:
  external/osqp:
    type: git
    url: https://github.com/oxfordcontrol/osqp
    version: master
  robotology/YARP:
    type: git
    url: https://github.com/robotology/yarp
    version: master
~~~
this would complexify its use as a reference for superbuild tags, while using a single `src` directory would result in a much more natural: 
~~~
repositories:
  osqp:
    type: git
    url: https://github.com/oxfordcontrol/osqp
    version: master
  YARP:
    type: git
    url: https://github.com/robotology/yarp
    version: master
~~~
this YAML file can also be used to specify a set of custom version for the robotology-superbuild itself, using the `ROBOTOLOGY_PROJECT_TAGS_CUSTOM_FILE` option, as described in https://github.com/robotology/robotology-superbuild/blob/master/doc/change-project-tags.md .

## Guidelines for affected users
If you are using the robotology-superbuild in [rolling update](https://github.com/robotology/robotology-superbuild#rolling-update) workflow  then pay attention that after you will update (i.e. `git pull`) the robotology-superbuild once https://github.com/robotology/robotology-superbuild#rolling-update is merged, your superbuild will use the `src` directory to store the repositories, instead of the `external` and `robotology`. To avoid confusion in which source files you will need to edit, it is suggested to either:
* do a **clean clone and build** of the superbuild
* or **manually delete** the `external`, `robotology`, `build/external` and `build/robotology` directories.

**Comment by traversaro** [comment]
The PR https://github.com/robotology/robotology-superbuild/pull/556 has been merged, so from now on if you using the `master` branch of the robotology-superbuild the changes described in this discussion will be effective! If you instead you are using the released version of the robotology-superbuild, you will need to wait for the next release to experience this change.

**Comment by S-Dafarra** [comment]
cc @robotology/iit-dynamic-interaction-control
